app/src/main/java/com/keylesspalace/tusky/util/LinkHelper.kt

```diff
@@ -0,0 +1,251 @@
+import com.keylesspalace.tusky.entity.HashTag
+import com.keylesspalace.tusky.entity.Status.Mention
+import com.keylesspalace.tusky.interfaces.LinkListener
+import java.net.URI
+import java.net.URISyntaxException
+
+fun getDomain(urlString: String?): String {
+    val uri: URI = try {
+        URI(urlString)
+    } catch (e: URISyntaxException) {
+        return ""
+    }
+
+    val host = uri.host
+    return when {
+        host == null -> ""
+        host.startsWith("www.") -> host.substring(4)
+        else -> host
+    }
+}
+
+/**
+ * Finds links, mentions, and hashtags in a piece of text and makes them clickable, associating
+ * them with callbacks to notify when they're clicked.
+ *
+ * @param view the returned text will be put in
+ * @param content containing text with mentions, links, or hashtags
+ * @param mentions any '@' mentions which are known to be in the content
+ * @param listener to notify about particular spans that are clicked
+ */
+fun setClickableText(view: TextView, content: CharSequence, mentions: List<Mention>?, tags: List<HashTag>?, listener: LinkListener) {
+    val builder = SpannableStringBuilder.valueOf(content)
+    val urlSpans = builder.getSpans(0, content.length, URLSpan::class.java)
+
+    for (span in urlSpans) {
+        setClickableText(span, builder, mentions, tags, listener)
+    }
+    view.text = builder
+    view.movementMethod = LinkMovementMethod.getInstance()
+}
+
+@VisibleForTesting
+fun setClickableText(
+    span: URLSpan,
+    builder: SpannableStringBuilder,
+    mentions: List<Mention>?,
+    tags: List<HashTag>?,
+    listener: LinkListener
+) {
+    val start = builder.getSpanStart(span)
+    val end = builder.getSpanEnd(span)
+    val flags = builder.getSpanFlags(span)
+    val text = builder.subSequence(start, end)
+
+    val customSpan = when {
+        text[0] == '#' -> getCustomSpanForTag(tags, span, listener)
+        text[0] == '@' -> getCustomSpanForMention(mentions, span, listener)
+        else -> null
+    } ?: object : NoUnderlineURLSpan(span.url) {
+        override fun onClick(view: View) {
+            listener.onViewUrl(url)
+        }
+    }
+
+    builder.removeSpan(span)
+    builder.setSpan(customSpan, start, end, flags)
+
+    /* Add zero-width space after links in end of line to fix its too large hitbox.
+     * See also : https://github.com/tuskyapp/Tusky/issues/846
+     *            https://github.com/tuskyapp/Tusky/pull/916 */
+    if (end >= builder.length || builder.subSequence(end, end + 1).toString() == "\n") {
+        builder.insert(end, "\u200B")
+    }
+}
+
+private fun getCustomSpanForTag(tags: List<HashTag>?, span: URLSpan, listener: LinkListener): ClickableSpan? {
+    return when (val tag = tags?.firstOrNull { it.url == span.url }) {
+        null -> null
+        else -> object : NoUnderlineURLSpan(span.url) {
+            override fun onClick(view: View) {
+                listener.onViewTag(tag.name)
+            }
+        }
+    }
```

Comment:
```suggestion
    return tags?.firstOrNull { it.url == span.url }?.let { tag ->
        object : NoUnderlineURLSpan(span.url) {
                        override fun onClick(view: View) = listener.onViewTag(tag.name)
        }
    }
```

(similar in other places)